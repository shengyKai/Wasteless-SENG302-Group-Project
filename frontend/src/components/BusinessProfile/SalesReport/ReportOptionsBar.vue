<template>
  <v-card color="secondary" dark>
    <v-card-actions>
      <v-row align="center" justify="center">
        <v-col cols="12" md="2" sm="12" class="mt-1">
          <v-menu
            ref="fromMenu"
            v-model="fromDateMenu"
            :return-value.sync="fromDate"
            :close-on-content-click="false"
            persistent
            offset-y
            min-width="auto"
          >
            <template v-slot:activator="{ on, attrs }">
              <v-text-field
                v-model="fromDate"
                label="From"
                prepend-inner-icon="mdi-calendar"
                readonly
                v-bind="attrs"
                v-on="on"
                outlined
              />
            </template>
            <v-date-picker
              v-model="fromDate"
              :max="maxFromDate"
              scrollable
            >
              <v-spacer/>
              <v-btn
                text
                color="primary"
                @click="fromDateMenu = false"
              >
                Cancel
              </v-btn>
              <v-btn
                text
                color="primary"
                @click="$refs.fromMenu.save(fromDate)"
              >
                OK
              </v-btn>
            </v-date-picker>
          </v-menu>
        </v-col>
        <v-col cols="12" md="2" sm="12">
          <v-menu
            class="mb-0 pb-0"
            ref="toMenu"
            v-model="toDateMenu"
            :return-value.sync="toDate"
            :close-on-content-click="false"
            persistent
            offset-y
            min-width="auto"
          >
            <template v-slot:activator="{ on, attrs }">
              <v-text-field
                v-model="toDate"
                label="To"
                prepend-inner-icon="mdi-calendar"
                readonly
                v-bind="attrs"
                v-on="on"
                outlined
              />
            </template>
            <v-date-picker
              v-model="toDate"
              :max="maxToDate"
              scrollable
            >
              <v-spacer/>
              <v-btn
                text
                color="primary"
                @click="toDateMenu = false"
              >
                Cancel
              </v-btn>
              <v-btn
                text
                color="primary"
                @click="$refs.toMenu.save(toDate)"
              >
                OK
              </v-btn>
            </v-date-picker>
          </v-menu>
        </v-col>
        <v-col cols="12" md="1" sm="12">
          <h4 class="font-weight-bold text-center">
            OR
          </h4>
        </v-col>
        <v-col cols="12" md="3" sm="12">
          <v-select
            v-model="periodBefore"
            flat
            solo-inverted
            hide-details
            label="Preset periods"
            :items="periodBeforeOptions"
            item-text="periodLevel"
            item-value="periodValue"
            prepend-inner-icon="mdi-clock-time-four"
            color="secondary"
          />
        </v-col>
        <v-divider vertical dark/>
        <v-col cols="12" md="2" sm="12">
          <v-select
            v-model="granularity"
            flat
            solo-inverted
            hide-details
            :items="granularityOptions"
            item-text="granularityLevel"
            item-value="granularityValue"
            color="secondary"
            label="Report Granularity"
          />
        </v-col>
        <v-col cols="12" md="1" sm="12" class="text-center">
          <v-btn
            color="primary"
            @click="sendReportSpecifications"
          >
            Generate
          </v-btn>
        </v-col>
      </v-row>
    </v-card-actions>
  </v-card>
</template>

<script>
export default {
  name: "ReportGenerationBar",
  data() {
    return {
      fromDate: null,
      toDate: null,
      fromDateMenu: false,
      toDateMenu: false,
      periodBefore: null,
      granularity: "yearly",
      periodBeforeOptions: [
        { periodLevel:'One day before', periodValue:'day' },
        { periodLevel:'One week before', periodValue:'week' },
        { periodLevel:'One month before', periodValue:'month' },
        { periodLevel:'One year before', periodValue:'year' }
      ],
      granularityOptions: [
        { granularityLevel:'Day', granularityValue:'daily' },
        { granularityLevel:'Week', granularityValue:'weekly' },
        { granularityLevel:'Month', granularityValue:'monthly' },
        { granularityLevel:'Year', granularityValue:'yearly' }
      ]
    };
  },
  computed: {
    /**
     * Max date of the fromDate has to be the same as the toDate or the present day
     */
    maxFromDate() {
      if (this.toDate !== null) {
        return this.toDate;
      } else {
        return new Date().toISOString().slice(0, 10);
      }
    },
    /**
     * Max date of the toDate is the present day
     */
    maxToDate() {
      return new Date().toISOString().slice(0, 10);
    }
  },
  methods: {
    /**
     * Method to send the report generation specifications to the parent, SalesReportPage
     */
    async sendReportSpecifications() {
      let [startDate, endDate] = this.getDatesForReport();
      this.$emit("sendRequestParams", {fromDate: startDate, toDate: endDate, granularity: this.granularity});
    },
    /**
     * Method to break down the dates from the report generation options so that it can provide logical values to the endpoint
     * ISSUE MAY BE HERE
     */
    getDatesForReport() {
      let startDate = new Date();
      let endDate = new Date();

      if (this.periodBefore === "day") {
        startDate.setDate( startDate.getDate() - 1);
      } else if (this.periodBefore === "week") {
        startDate.setDate( startDate.getDate() - 6);
      } else if (this.periodBefore === "month") {
        startDate.setMonth( startDate.getMonth() - 1);
      } else if (this.periodBefore === "year") {
        startDate.setFullYear( startDate.getFullYear() - 1);
      }

      startDate = startDate.toISOString().slice(0, 10);
      endDate = endDate.toISOString().slice(0, 10);

      if (this.fromDate !== null && this.toDate !== null) {
        startDate = this.fromDate;
        endDate = this.toDate;
      }
      return [startDate, endDate];
    }
  },
  watch: {
    /**
     * If the fromDate value is null and a value is being set for the toDate, the fromDate value is also set to the toDate value.
     * This is to ensure both fields are filled in at all times.
     * If toDate is not null, periodBefore must be null as both options cannot be used at the same time.
     * Else if toDate is null and fromDate is not null, make fromDate null. This is to ensure both dates are filled in or not at all
     * times.
     * If the toDate date is before the fromDate date, change the fromDate to that value
     */
    toDate(value) {
      if (this.fromDate === null) {
        this.fromDate = value;
      }
      if (value !== null) {
        this.periodBefore = null;
      } else if (value === null && this.fromDate !== null) {
        this.fromDate = null;
      }
      if (new Date(value) < new Date(this.fromDate)) {
        this.fromDate = value;
      }
    },
    /**
     * If the toDate value is null and a value is being set for the fromDate, the toDate value is also set to the fromDate value.
     * This is to ensure both fields are filled in at all times.
     * If fromDate is not null, periodBefore must be null as both options cannot be used at the same time.
     * Else if fromDate is null and toDate is not null, make toDate null. This is to ensure both dates are filled in or not at all
     * times.
     * If the fromDate date is after the toDate date, change the toDate to that value
     * Technically this situation should not have a chance to occur, but acts as a sanity check
     */
    fromDate(value) {
      if (this.toDate === null) {
        this.toDate = value;
      }
      if (value !== null) {
        this.periodBefore = null;
      } else if (value === null && this.toDate !== null) {
        this.toDate = null;
      }
      if (new Date(value) > new Date(this.toDate)) {
        this.toDate = value;
      }
    },
    /**
     * This watches the periodBefore variable so that if the period is not null, the toDate and fromDate
     * must be null, as both options cannot be used at the same time.
     */
    periodBefore(value) {
      if (value !== null) {
        this.toDate = null;
        this.fromDate = null;
      }
    }
  }
};
</script>

<style scoped>
/*
  The bottom two css styles had to be done because by Vueitfy's default implementation, there are sometimes some components
  which are created which causes the layout to be unorganised. The >>> symbol is a deep selector symbol which targets the specific
  Vue/Vueitfy class when you provide a parent class. Try uncommenting the bottom two styles to see the difference
 */
.v-input >>> .v-text-field__details{
  margin-bottom: 0;
  display: none;
}
.v-input >>> .v-input__slot{
  margin-bottom: 0;
}
</style>