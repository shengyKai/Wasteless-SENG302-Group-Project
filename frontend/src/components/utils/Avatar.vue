<template>
  <v-avatar
    :size="sizeInPixels"
    :color="user !== undefined ? 'primary' : 'secondary'"
    class="white--text headline"
    :style="`font-size: ${fontSize}px !important`"
  >
    <template v-if="image === undefined">{{ initials }}</template>
    <v-img v-else :src="image" height="100%"/>
  </v-avatar>
</template>

<script>
import { imageSrcFromFilename } from '@/utils';

export default {
  name: 'Avatar',
  props: {
    /**
     * User to make avatar for.
     * Could be aquired from the 'getUser' function in api.ts
     */
    user: Object,

    /**
     * Business to make avatar for.
     */
    business: Object,


    /**
     * The size of the avatar
     * The small and medium sizes use the thumbnail sized avatar image, large uses full size.
     * @values small, medium, large
     */
    size: {
      validator: (s) => ['small', 'medium', 'large'].includes(s),
      default: 'medium',
    },
  },

  computed: {
    sizeInPixels() {
      switch (this.size) {
      case "small":
        return 30;
      case "medium":
        return 40;
      case "large":
        return 200;
      default:
        throw new Error(`Invalid size "${this.size}"`);
      }
    },
    fontSize() {
      return this.sizeInPixels * 0.5;
    },
    initials() {
      if (this.user !== undefined) {
        return this.user.firstName[0].toUpperCase() + this.user.lastName[0].toUpperCase();
      } else if (this.business !== undefined) {
        let pieces = this.business.name.split(/\s/);
        if (pieces.length === 2) {
          return pieces[0][0].toUpperCase() + pieces[1][0].toUpperCase();
        }
        return this.business.name[0].toUpperCase();
      } else {
        throw new Error('Either "user" or "business" must be provided.');
      }
    },
    image() {
      let image;
      if (this.user !== undefined) {
        image = this.user.images[0];
      } else if (this.business !== undefined) {
        image = this.business.images[0];
      } else {
        throw new Error('Either "user" or "business" must be provided.');
      }

      if (image === undefined) return undefined;

      let url = image.thumbnailFilename;
      if (url === undefined || url === null) {
        url = image.filename;
      }
      return imageSrcFromFilename(url);
    }
  }
};
</script>

<style scoped>
.profile-image {
  margin: auto;
  position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
}
</style>