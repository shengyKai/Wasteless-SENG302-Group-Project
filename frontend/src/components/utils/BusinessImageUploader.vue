<template>
  <v-dialog v-model="showDialog" persistent max-width="500px">
    <v-form>
      <v-card>
        <v-card-title>
          Upload Business Image
        </v-card-title>
        <v-card-text>
          <ImageSelector v-model="file"/>
        </v-card-text>
        <v-card-actions>
          <v-spacer/>
          <!-- Current error message -->
          <div class="error-text" v-if="errorMessage !== undefined"> {{ errorMessage }} </div>
          <v-spacer/>
          <v-btn
            color="primary"
            text
            @click="closeForm">
            Close
          </v-btn>
          <v-btn
            type="submit"
            color="primary"
            :disabled="!file"
            :loading="isLoading"
            @click.prevent="uploadImage">
            Upload
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-form>
  </v-dialog>
</template>

<script>
import ImageSelector from "@/components/utils/ImageSelector";

export default {
  name: "BusinessImageUploader",
  components: {ImageSelector},
  props: {
    value: undefined,
  },
  data() {
    return {
      isLoading: false,
      errorMessage: undefined,
      showDialog: true
    };
  },
  methods: {
    /*
     * Close the dialog and emit message to the parent to add file to uploaded images.
     */
    uploadImage() {
      this.$emit('uploadImage');
    },
    /**
     * Closes the dialog and clears any selected files
     */
    closeForm() {
      this.file = undefined;
      this.$emit('closeDialog');
    },
  },
  computed: {
    file: {
      get() {
        return this.value;
      },
      set (value) {
        this.$emit('input', value);
      },
    }
  }
};
</script>