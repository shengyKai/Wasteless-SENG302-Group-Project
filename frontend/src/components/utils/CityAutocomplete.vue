<template>
  <v-combobox
    label="City"
    :items="cityItems"
    :loading="isLoading"
    :search-input.sync="citySearch"
    :rules="mandatoryRules.concat(maxCharRule)"
    no-filter
    clearable
    outlined
  />
</template>

<script>
export default {
  name: 'CityAutocomplete',
  data () {
    return {
      cityItems: [],
      isLoading: false,
      citySearch: null,
      mandatoryRules: [
        //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
        //if it does not follow the format, display error message
        field => !!field || 'Field is required'
      ],
      maxCharRule: [
        field => (field.length <= 100) || 'Reached max character limit: 100'
      ]
    };
  },

  watch: {
    citySearch (val) {
      this.cityItems = [];

      //if input length exists, and has a length more than 2
      if (val && val.length > 2) {
        //show loading animation
        this.isLoading = true;
        //append to the api url the input the user has entered
        let url = `https://photon.komoot.io/api/?q=${encodeURIComponent(val)}&osm_tag=place:city&osm_tag=place:town`;

        fetch(url).then(res => res.json()).then(res => {
          res.features.forEach(feature => {
            //If the returned GEOJSON has any of these key(s):
            //city
            //it will add that result into the autocomplete suggestion
            this.cityItems.push(feature.properties.name);
          });
        }).catch(err => {
          console.log(err);
          //after everything is shown, the loading animation will stop
        }).finally(() => (this.isLoading = false));
      }
    },
  }
};
</script>
