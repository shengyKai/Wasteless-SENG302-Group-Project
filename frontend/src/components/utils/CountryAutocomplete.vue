<template>
  <v-combobox
    label="Country"
    :items="countryItems"
    :loading="isLoading"
    :search-input.sync="countrySearch"
    :rules="mandatoryRules.concat(maxCharRule)"
    no-filter
    clearable
    outlined
  />
</template>

<script>
export default {
  name: 'CountryAutocomplete',
  data () {
    return {
      countryItems: [],
      isLoading: false,
      countrySearch: null,
      mandatoryRules: [
        //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
        //if it does not follow the format, display error message
        field => !!field || 'Field is required'
      ],
      maxCharRule: [
        field => (field.length <= 100) || 'Reached max character limit: 100'
      ]
    };
  },

  watch: {
    countrySearch (val) {
      this.countryItems = [];

      //if input length exists, and has a length more than 2
      if (val && val.length > 2) {
        //show loading animation
        this.isLoading = true;
        //append to the api url the input the user has entered
        let url = `https://photon.komoot.io/api/?q=${encodeURIComponent(val)}&osm_tag=place:country`;

        fetch(url).then(res => res.json()).then(res => {
          res.features.forEach(feature => {
            //If the returned GEOJSON has any of these key(s):
            //country
            //it will add that result into the autocomplete suggestion
            this.countryItems.push(feature.properties.name);
          });
        }).catch(err => {
          console.log(err);
          //after everything is shown, the loading animation will stop
        }).finally(() => (this.isLoading = false));
      }
    },
  }
};
</script>
