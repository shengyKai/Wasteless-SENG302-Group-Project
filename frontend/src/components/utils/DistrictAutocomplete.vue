<template>
  <v-combobox
      v-model="district"
      label="District/City Area"
      :items="districtItems"
      :loading="isLoading"
      :search-input.sync="districtSearch"
      no-filter
      clearable
      outlined
  />
</template>

<script>
export default {
  name: 'DistrictAutocomplete',
  data () {
    return {
      district: '',
      districtItems: [],
      isLoading: false,
      districtSearch: null
    }
  },

  watch: {
    districtSearch (val) {
      this.districtItems = []

      //if input length exists, and has a length more than 2
      if (val && val.length > 2) {
        //show loading animation
        this.isLoading = true
        //append to the api url the input the user has entered
        let url = 'https://photon.komoot.io/api/?q=' + val

        fetch(url).then(res => res.json()).then(res => {
          res.features.forEach(feature => {
            //If the returned GEOJSON has any of these key(s):
            //district
            //it will add that result into the autocomplete suggestion
            if (feature.properties.district !== undefined) {
              this.districtItems.push(feature.properties.district)
            }
          })
        }).catch(err => {
          console.log(err)
          //after everything is shown, the loading animation will stop
        }).finally(() => (this.isLoading = false))
      }
    },
  }
}
</script>
