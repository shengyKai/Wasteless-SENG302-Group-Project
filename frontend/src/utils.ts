import { MarketplaceCardSection } from "./api/internal";

/**
 * Sets a cookie. Will expire in one year.
 * @param name Name of the cookie being set.
 * @param value Value to store in the cookie.
 */
export function setCookie(name: string, value: string|number) {
  const date = new Date();
  date.setFullYear(date.getFullYear() + 1);
  document.cookie = `${name}=${value}; expires=${date.toUTCString()}; path=/`;
}

/**
 * Reads a cookie.
 * @param name Name of the cookie to read.
 * @returns Value of the cookie. NULL if the cookie does not exist.
 */
export function getCookie(name: string) {
  let cookies = document.cookie.split(';');
  for (let cookie of cookies) {
    cookie = cookie.trim();
    if (cookie.startsWith(name + '=')) {
      return cookie;
    }
  }
  return null;
}

/**
 * Deletes a cookie.
 * @param name Name of the cookie to delete.
 */
export function deleteCookie(name: string) {
  document.cookie = `${name}=; Path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
}

/**
 * Cookie names.
 */
export const COOKIE = {
  USER: 'user'
};

/**
 * Converts a date to a human readable string
 *
 * @param date Date to format, either a date object or a parsable date string
 * @returns Date in a human readable string
 */
export function formatDate(date: Date | string) {
  if (typeof date === 'string') date = new Date(date);
  const parts = date.toDateString().split(' ');
  return `${parts[2]} ${parts[1]} ${parts[3]}`;
}

export function formatPrice(price : number) {
  if (Number.isInteger(price)) {
    return price.toString();
  } else {
    return (+price).toFixed(2);
  }
}

/**
 * User roles.
 */
export const USER_ROLES = {
  USER: 'user',
  GAA: 'globalApplicationAdmin',
  DGAA: 'defaultGlobalApplicationAdmin'
};

/**
 * Mapping between section key and section name
 */
export const SECTION_NAMES: Record<MarketplaceCardSection, string> = {
  ForSale: 'For Sale',
  Wanted: 'Wanted',
  Exchange: 'Exchange',
};


/**
 * Rate limits the input function by waiting a given amount of time calling the inner function.
 * If interrupted with another call then this wait is reset.
 *
 * @param func Function to rate limit
 * @param wait Time (ms) to wait before calling the function
 */
export function debounce(func: (() => void), wait: number) {
  let timeout: number | undefined;
  function debounced() {
    if (timeout !== undefined) {
      clearTimeout(timeout);
    }
    timeout = setTimeout(func, wait);
  }
  return debounced;
}

/**
 * Trims the input string such that it is always less than or equal to the provided length
 * This function will try to avoid breaking up words unless there is no alternative
 * @param str String to trim
 * @param length Maximum length
 * @returns Trimmed string
 */
export function trimToLength(str: string, length: number) {
  const re = new RegExp(`^(.{0,${length}})(\\s.*|$)`);
  return str.replace(re, "$1").slice(0, length);
}

/**
 * Checks whether we are executing in the test enviroment
 *
 * @returns true if we're testing false otherwise
 */
export function isTesting() {
  return process.env.JEST_WORKER_ID !== undefined;
}

/**
 * ========= REGEX =========
 */

/**
 * Returns a Regex pattern that matches letters, spaces, dashes, hyphen and apostrophe
 */
export function regxAlphabet() {
  return /^[\p{L}\-.' ]*$/u;
}

/**
 * Returns a Regex pattern that matches letters, spaces, numbers and common characters
 */
export function regxAlphabetExtended() {
  return /(^[\p{L}\d\p{P} ]*$)/u;
}

export function regxAlphabetExtendedMultiline() {
  return /(^[\p{L}\d\p{P}$\s]*$)/u;
}

/**
 * Returns a Regex pattern that matches numbers
 */
export function regxNumerical() {
  return /(^[\d]*$)/;
}

/**
 * Returns a Regex pattern that matches a price
 */
export function regxPrice() {
  return /(^\d{1,8}(\.\d{2})?$)|^$/;
}

/**
 * Returns a Regex that matches a password with at least one number and one letter.
 */
export function regxPassword() {
  return /^(?=.*[0-9])(?=.*[\p{L} ])([\p{L}0-9 !@#$%^&*()-]+)$/u;
}

/**
 * Returns a Regex that matches a valid email address
 */
export function regxEmail() {
  return /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
}

/**
 * Returns a Regex that matches a valid post code
 */
export function regxPostCode() {
  return /(^[\p{L}0-9]*$)/u;
}

/**
 Returns a Regex that matches a valid phone number
 */
export function regxPhoneNumber() {
  return /(^\(?\d{1,3}\)?[\s.-]?\d{3,4}[\s.-]?\d{4,5}$)|(^$)/;
}

/**
 Returns a Regex that matches a country code
 */
export function regxCountryCode() {
  return /(^(\d{1,2}-)?\d{2}$)|(^$)/;
}
/**
 Returns a Regex that matches a valid street address
 */
export function regxStreet() {
  return /^(([0-9]+|[0-9]+\/[0-9]+)[\p{L}]?)(?=\s[\p{L}])([\p{L}0-9-.' ]+)$/u;
}
/**
 Returns a Regex that matches a valid product code
 */
export function regxProductCode() {
  return /^[-A-Z0-9]+$/;
}

/**
 * ========= RULES =========
 */

export const emailRules = [
  (email: string) => regxEmail().test(email) || 'E-mail must be valid'
];

export const mandatoryRules = [
  //All fields with the class "required" will go through this ruleset to ensure the field is not empty.
  //if it does not follow the format, display error message
  (field: string) => !!field || 'Field is required'
];

export const passwordRules = [
  (field: string) => (field && field.length >= 7) || 'Password must have 7+ characters',
  (field: string) => regxPassword().test(field) || 'Must have at least one number and one letter'
];

export const postCodeRules = [
  (field: string) => regxPostCode().test(field) || 'Must contain numbers and alphabet only'
];

export const nameRules = [
  (field: string) =>  (field.length === 0 || regxAlphabet().test(field)) || 'Naming must only contain letter and common characters'
];

export function maxCharRules(size: number) {
  return [
    (field: string) => (field.length <= size) || `Reached max character limit: ${size}`
  ];
}

export const phoneNumberRules = [
  (field: string) => regxPhoneNumber().test(field) || 'Must only contain 8-12 digits'
];
export const countryCodeRules = [
  (field: string) => regxCountryCode().test(field) || 'Must only contain 2 digits.'
];
export const alphabetRules = [
  (field: string) => (field.length === 0 || regxAlphabet().test(field)) || 'Naming must be valid'
];
export const streetNumRules = [
  (field: string) => (field && field.length <= 109) || 'Reached max character limit 109 ',
  (field: string) => regxStreet().test(field) || 'Must contain unit number and street name'
];

export const quantityRules = [
  (field: string) => regxNumerical().test(field) || 'Must contain an integer',
  (field: string) => field ? parseInt(field) > 0 || 'Must be greater than zero' : true
];

export const smallPriceRules = [
  //A price must be numbers and may contain a decimal followed by exactly two numbers (4digit)
  (field: string) => regxPrice().test(field) || 'Must be a valid price',
  (field: string) => field ? parseInt(field) < 10000 || 'Must be less than 10,000' : true,
];

export const hugePriceRules = [
  //A price must be numbers and may contain a decimal followed by exactly two numbers (6digit)
  (field: string) => regxPrice().test(field) || 'Must be a valid price',
  (field: string) => field ? parseInt(field) < 1000000 || 'Must be less than 1,000,000' : true
];

export const alphabetExtendedSingleLineRules = [
  (field: string) => regxAlphabetExtended().test(field) || 'Must only contain letters, numbers, punctuation and spaces',
];

export const alphabetExtendedMultilineRules = [
  (field: string) => regxAlphabetExtendedMultiline().test(field) || 'Must only contain letters, numbers, punctuation and whitespace',
];

export const productCodeRules = [
  (field: string) => field.length <= 15 || 'Reached max character limit: 15',
  (field: string) => !/ /.test(field) || 'Must not contain a space',
  (field: string) => regxProductCode().test(field) || 'Must be all uppercase letters, numbers and dashes.',
];
