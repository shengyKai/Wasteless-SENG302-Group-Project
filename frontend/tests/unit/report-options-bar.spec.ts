import Vue from 'vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import ReportOptionsBar from "@/components/BusinessProfile/SalesReport/ReportOptionsBar.vue";
import {findButtonWithText} from "./utils";

Vue.use(Vuetify);

const currentDate = new Date("2012-12-21");
describe('ReportOptionsBar.vue', () => {
  let wrapper: Wrapper<any>;
  let vuetify: Vuetify;


  function findButton(component: string) {
    return findButtonWithText(wrapper, component);
  }

  beforeAll(() => {
    jest
      .useFakeTimers('modern')
      .setSystemTime(currentDate);
  });


  beforeEach(async () => {
    const localVue = createLocalVue();
    vuetify = new Vuetify();

    wrapper = mount(ReportOptionsBar, {
      localVue,
      vuetify,
      data() {
        return {
          fromDate: new Date("2012-01-01").toISOString().slice(0, 10),
          toDate: new Date("2012-02-01").toISOString().slice(0, 10),
          granularity: "yearly",
        };
      }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it("The maximum/minimum possible date for the From and To datepickers are correct", ()=> {
    expect(new Date(wrapper.vm.maxFromDate)).toStrictEqual(new Date("2012-02-01")); // Maximum fromDate is toDate
    expect(new Date(wrapper.vm.maxToDate)).toStrictEqual(currentDate); // Maximum toDate is now
    expect(new Date(wrapper.vm.minToDate)).toStrictEqual(new Date("2012-01-01")); // Minimum toDate is fromDate
  });

  it("If toDate is not set, maxFromDate will be the present day", async() => {
    await wrapper.setData({
      toDate: undefined,
    });
    expect(new Date(wrapper.vm.maxFromDate)).toStrictEqual(currentDate);
  });

  it('If preset period "One year before" is selected, then the fromDate/toDate should be set accordingly', async () => {
    await wrapper.setData({
      presetPeriodUserString: 'One year before',
    });

    expect(new Date(wrapper.vm.fromDate)).toStrictEqual(new Date("2011-12-21"));
    expect(new Date(wrapper.vm.toDate)).toStrictEqual(currentDate);
  });

  it('If preset period "One month before" is selected, then the fromDate/toDate should be set accordingly', async () => {
    await wrapper.setData({
      presetPeriodUserString: 'One month before',
    });

    expect(new Date(wrapper.vm.fromDate)).toStrictEqual(new Date("2012-11-21"));
    expect(new Date(wrapper.vm.toDate)).toStrictEqual(currentDate);
  });

  it('If preset period "One day before" is selected, then the fromDate/toDate should be set accordingly', async () => {
    await wrapper.setData({
      presetPeriodUserString: 'One day before',
    });

    expect(new Date(wrapper.vm.fromDate)).toStrictEqual(currentDate);
    expect(new Date(wrapper.vm.toDate)).toStrictEqual(currentDate);
  });

  it("Clicking on generate will emit an event to SalesReportPage along with the specified report options", async () => {
    await findButton("Generate").trigger("click");
    expect(wrapper.emitted().sendRequestParams).toBeTruthy();
    expect(wrapper.emitted().sendRequestParams as any[1]).toEqual([[{
      fromDate: new Date("2012-01-01").toISOString().slice(0, 10),
      toDate: new Date("2012-02-01").toISOString().slice(0, 10),
      granularity: "yearly"
    }]]);
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});