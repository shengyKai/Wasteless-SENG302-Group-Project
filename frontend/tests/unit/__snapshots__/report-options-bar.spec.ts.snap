// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportOptionsBar.vue Matches snapshot 1`] = `
<div class="v-card v-sheet theme--dark secondary">
  <div class="v-card__actions">
    <div class="row align-center justify-center">
      <div class="col-md-2 col-12">
        <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--dark v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-input__prepend-inner">
                <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--dark"></i></div>
              </div>
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-136" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">From</label><input role="button" aria-haspopup="true" aria-expanded="false" id="input-136" readonly="readonly" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
        <div class="v-menu" persistent="">
          <!---->
        </div>
      </div>
      <div class="col-md-2 col-12">
        <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--dark v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-input__prepend-inner">
                <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--dark"></i></div>
              </div>
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-141" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">To</label><input role="button" aria-haspopup="true" aria-expanded="false" id="input-141" readonly="readonly" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
        <div class="v-menu mb-0 pb-0" persistent="">
          <!---->
        </div>
      </div>
      <div class="col-md-3 col-12">
        <div class="v-input v-input--hide-details theme--dark v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--solo-flat v-text-field--enclosed v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-145" class="v-input__slot">
              <div class="v-input__prepend-inner">
                <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate mdi mdi-clock-time-four theme--dark"></i></div>
              </div>
              <div class="v-select__slot"><label for="input-145" class="v-label theme--dark" style="left: 0px; position: absolute;">Preset periods</label>
                <div class="v-select__selections"><input id="input-145" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                </div><input type="hidden">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <hr role="separator" aria-orientation="vertical" class="v-divider v-divider--vertical theme--dark">
      <div class="col-md-2 col-12">
        <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--dark v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--solo-flat v-text-field--enclosed v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-150" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">Year</div><input id="input-150" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--dark"></i></div>
                </div><input type="hidden" value="yearly">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="text-center col-md-2 col-12"><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--default primary"><span class="v-btn__content">
          Generate
        </span></button></div>
    </div>
  </div>
</div>
`;
